### ISSUE TITLE
needsrestart perl warnings under perl 5.20

### ISSUE THREAD

**Issue by glensc** [liske/needrestart#6]
it prints too many notices here and there....

```
perl --version|head -n2

This is perl 5, version 20, subversion 1 (v5.20.1) built for x86_64-pld-linux-thread-multi
```

tried to see what options it supports:
```
# needrestart -h
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Unknown option: h
Usage:

  needrestart [-vn] [-c <cfg>] [-r <mode>] [-f <fe>] [-bkl]
```

tried to see what is it's version:

```
# needrestart -v
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
[main] eval /etc/needrestart/needrestart.conf
[main] running in root-mode
[Core] Using UI 'NeedRestart::UI::stdio'...
[main] detected systemd
Argument "" isn't numeric in sort at /usr/sbin/needrestart line 276, <> line 1.
[Core] #2377 is a NeedRestart::Interp::Perl
[Core] #2792 is a NeedRestart::Interp::Python
[Core] #2810 is a NeedRestart::Interp::Python
[main] #5246 uses non-existing /home/glen/.cache/IntelliJIdea/tmp/jna/jna6470877052996353095.tmp
[main] #5246 is a child of #5211
[main] #5211 exe => /bin/mksh
[main] #5211 unexpected cgroup ''
[main] #5211 running /etc/needrestart/hook.d/20-rpm
[main] #5211 running /etc/needrestart/hook.d/90-none
[Kernel] Linux: kernel release 3.14.36-1, kernel version #1 SMP Fri Mar 20 10:29:45 CET 2015
Use of uninitialized value in numeric lt (<) at /usr/share/perl5/vendor_perl/NeedRestart/Kernel/Linux.pm line 97.
[Kernel/Linux] /boot/vmlinuz.old seems to be too small
[Kernel/Linux] /boot/vmlinuz-3.14.36-1 => 3.14.36-1 (builder@ymir-builder) #1 SMP Fri Mar 20 10:29:45 CET 2015 [3.14.36-1]*
[Kernel/Linux] /boot/vmlinuz-3.14.34-1 => 3.14.34-1 (builder@ymir-builder) #1 SMP Mon Mar 2 00:39:35 CET 2015 [3.14.34-1]
[Kernel/Linux] /boot/vmlinuz-3.6.10-4.fc18.x86_64 => 3.6.10-4.fc18.x86_64 (mockbuild@) #1 SMP Tue Dec 11 18:01:27 UTC 2012 [3.6.10-4.fc18.x86_64]
Use of uninitialized value in numeric lt (<) at /usr/share/perl5/vendor_perl/NeedRestart/Kernel/Linux.pm line 97.
[Kernel/Linux] /boot/vmlinuz seems to be too small
[Kernel/Linux] Expected linux version: 3.14.36-1
Running kernel seems to be up-to-date.
No services need to be restarted.
```

so the actual run:
```
# needrestart -n -r l 
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Argument "" isn't numeric in sort at /usr/sbin/needrestart line 276, <> line 1.
Use of uninitialized value in numeric lt (<) at /usr/share/perl5/vendor_perl/NeedRestart/Kernel/Linux.pm line 97.
Use of uninitialized value in numeric lt (<) at /usr/share/perl5/vendor_perl/NeedRestart/Kernel/Linux.pm line 97.
Running kernel seems to be up-to-date.
No services need to be restarted.
```

let me know if you need more info!

**Comment by glensc** [comment]
some external module versions:

perl-Module-Find-0.12-1.noarch
perl-Module-ScanDeps-1.10-1.noarch
perl-Proc-ProcessTable-0.48-3.x86_64
perl-Sort-Naturally-1.02-2.noarch
perl-Term-ReadKey-2.31-3.x86_64

**Comment by liske** [comment]
Hi,

perl 5.20 should work - running needrestart on Debian jessie (perl 5.20.2) is not that noisy.

```console
# needrestart -h
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
```

It seems to be triggered by a bug in Proc::ProcessTable. It was fixed some time ago: jwbargsten/perl-proc-processtable@c4d07f803ca37dc6afc0b414b6b047c7b2fb813c

```
Argument "" isn't numeric in sort at /usr/sbin/needrestart line 276, <> line 1.
```
Which version of needrestart did you try (git HEAD / github release)? Btw: only longopts --help/--version is supported.

```
Use of uninitialized value in numeric lt (<) at /usr/share/perl5/vendor_perl/NeedRestart/Kernel/Linux.pm line 97.
[Kernel/Linux] /boot/vmlinuz.old seems to be too small
[..]
Use of uninitialized value in numeric lt (<) at /usr/share/perl5/vendor_perl/NeedRestart/Kernel/Linux.pm line 97.
[Kernel/Linux] /boot/vmlinuz seems to be too small
```
The warning during kernel probing seems to be triggered by a dangling symlink (/boot/vmlinuz.old), isn't it?

**Comment by glensc** [comment]
for perl-proc-processtable@c4d07f8 maybe you bump your `use Proc::ProcessTable X.Y` to require newer fixed version?

i am using 2.0 release tarball from github

yes, both `/boot/vmlinuz.old` and `/boot/vmlinuz` are dead symlinks. is the prior warning from the same issue?

for short `-h` i just guessed, program will print usage or just "uknown option error" and abort :)

**Comment by glensc** [comment]
now with 601ab27:

```
# needrestart -n -r l 
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Ran into unknown state (hex char: 29) at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Use of uninitialized value in list assignment at /usr/share/perl5/vendor_perl/NeedRestart/Utils.pm line 47.
Argument "" isn't numeric in sort at /usr/sbin/needrestart line 305, <> line 1.                                              ]
Scanning processes...                                                                                                          
Scanning candidates...                                                                                                         
Scanning linux images...                                                                                                       
Running kernel seems to be up-to-date.
No services need to be restarted.
```

**Comment by liske** [comment]
Thanks for giving 601ab27 a try. The warning in line 305 is related to perl-proc-processtable@c4d07f8.

Using the `use Proc::ProcessTable X.Y` statement should only be used for features missing in older versions of a package IMHO - bug fixes might be backported.

Ignoring unknown argv parameters is required while using the debconf frontend on Debian GNU/Linux systems.

**Comment by glensc** [comment]
Do you know what is the version that is required for the Proc::ProcessTable bug to disappear? for the sake when somebody else gets this error and reaches this ticket.

**Comment by liske** [comment]
Seems to be fixed in Proc::ProcessTable 0.49+ (jwbargsten/perl-proc-processtable@1dc2ddceac4fa2b2ac7881c71d7bbd1cf5da846a)
